Documentation/man/man3/targeterImage.3
.TH "targeterImage" 3 "Fri Mar 17 2017" "Version 1" "targeter" \" -*- nroff -*-
.ad l
.nh
.SH NAME
targeterImage
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <targeterimage\&.h>\fP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBtargeterImage\fP (\fBImagesContainer\fP *pWin)"
.br
.ti -1c
.RI "\fBtargeterImage\fP ()"
.br
.ti -1c
.RI "\fB~targeterImage\fP ()"
.br
.ti -1c
.RI "void \fBaddImageFromFile\fP (std::string filename, imageType::imageType type=imageType::display)"
.br
.ti -1c
.RI "void \fBaddImageEx\fP (cv::Mat im, imageType::imageType type, std::string descriptiveName='', std::string tooltip='', std::string filename='')"
.br
.ti -1c
.RI "void \fBaddImage\fP (cv::Mat im)"
.br
.ti -1c
.RI "cv::Mat & \fBgetImage\fP ()"
.br
.ti -1c
.RI "void \fBcreateQTImage\fP (bool bRGBSwap=true)"
.br
.ti -1c
.RI "QImage \fBgetQTImage\fP ()"
.br
.ti -1c
.RI "std::string \fBsetDefaultName\fP ()"
.br
.ti -1c
.RI "void \fBsetImageType\fP (imageType::imageType type)"
.br
.ti -1c
.RI "imageType::imageType \fBgetImageFunction\fP ()"
.br
.ti -1c
.RI "std::vector< int > & \fBgetDerivedImageArray\fP ()"
.br
.ti -1c
.RI "void \fBaddDerivedIndex\fP (int ind)"
.br
.ti -1c
.RI "void \fBaddDerivedIndexes\fP (const std::vector< int > &to_add)"
.br
.ti -1c
.RI "\fBtargeterImage\fP & \fBgetImageOfType\fP (imageType::imageType imageFunction)"
.br
.ti -1c
.RI "void \fBremoveDerivedIndex\fP (int ind)"
.br
.ti -1c
.RI "void \fBremoveDerivedIndexes\fP (const std::vector< int > &to_remove)"
.br
.ti -1c
.RI "std::string \fBgetName\fP ()"
.br
.ti -1c
.RI "void \fBsetName\fP (std::string s)"
.br
.ti -1c
.RI "int * \fBget1DImage\fP (imageType::imageType type)"
.br
.ti -1c
.RI "void \fBset1DImage\fP (int *im, imageType::imageType type=imageType::display)"
.br
.ti -1c
.RI "void \fBfree1DImages\fP ()"
.br
.ti -1c
.RI "int \fBgetImageRows\fP ()"
.br
.ti -1c
.RI "int \fBgetImageCols\fP ()"
.br
.ti -1c
.RI "cv::Size \fBgetImageSize\fP ()"
.br
.ti -1c
.RI "int \fBgetImageType\fP ()"
.br
.ti -1c
.RI "cv::Mat & \fBgetLabelsImage\fP ()"
.br
.ti -1c
.RI "cv::Mat & \fBgetStatsImage\fP ()"
.br
.ti -1c
.RI "cv::Mat & \fBgetCentroidsImage\fP ()"
.br
.ti -1c
.RI "void \fBsetLabelsImage\fP (cv::Mat im)"
.br
.ti -1c
.RI "void \fBsetStatsImage\fP (cv::Mat im)"
.br
.ti -1c
.RI "void \fBsetCentroidsImage\fP (cv::Mat im)"
.br
.ti -1c
.RI "void \fBsetMaskType\fP (drawingMode::drawingMode mode)"
.br
.ti -1c
.RI "drawingMode::drawingMode \fBgetMaskType\fP ()"
.br
.in -1c
.SH "Detailed Description"
.PP 
class to represent targeter image used to store associated image information and group images together 
.PP
Definition at line 16 of file targeterimage\&.h\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "targeterImage::targeterImage (\fBImagesContainer\fP * pImages)"
Constructor for \fBtargeterImage\fP class, has pointer to image array
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/17
.RE
.PP
FullName \fBtargeterImage::targeterImage\fP Qualifier 
.PP
\fBParameters:\fP
.RS 4
\fI\fBImagesContainer\fP\fP * pImages 
.RE
.PP
\fBReturns:\fP
.RS 4
Access public 
.RE
.PP

.PP
Definition at line 26 of file targeterimage\&.cpp\&.
.SS "targeterImage::targeterImage ()"
Constructor for \fBtargeterImage\fP class
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/17
.RE
.PP
FullName \fBtargeterImage::targeterImage\fP Qualifier 
.PP
\fBReturns:\fP
.RS 4
Access public 
.RE
.PP

.PP
Definition at line 50 of file targeterimage\&.cpp\&.
.SS "targeterImage::~targeterImage ()"
\fBtargeterImage\fP destructor
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/17
.RE
.PP
FullName \fBtargeterImage::~targeterImage\fP Qualifier 
.PP
\fBReturns:\fP
.RS 4
Access public 
.RE
.PP

.PP
Definition at line 73 of file targeterimage\&.cpp\&.
.SH "Member Function Documentation"
.PP 
.SS "void targeterImage::addImage (cv::Mat im)"
adds image and creates QT Image copy for display
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/17
.RE
.PP
FullName \fBtargeterImage::addImage\fP Qualifier 
.PP
\fBParameters:\fP
.RS 4
\fIcv::Mat\fP im 
.RE
.PP
\fBReturns:\fP
.RS 4
void Access public 
.RE
.PP

.PP
Definition at line 345 of file targeterimage\&.cpp\&.
.SS "void targeterImage::addImageEx (cv::Mat im, imageType::imageType type, std::string descriptiveName = \fC''\fP, std::string tooltip = \fC''\fP, std::string filename = \fC''\fP)"
Extended method to create image
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/17
.RE
.PP
FullName \fBtargeterImage::addImageEx\fP Qualifier 
.PP
\fBParameters:\fP
.RS 4
\fIcv::Mat\fP im 
.br
\fIimageType::imageType\fP type 
.br
\fIstd::string\fP descriptiveName 
.br
\fIstd::string\fP tooltip 
.br
\fIstd::string\fP filename 
.RE
.PP
\fBReturns:\fP
.RS 4
void Access public 
.RE
.PP

.PP
Definition at line 282 of file targeterimage\&.cpp\&.
.SS "void targeterImage::addImageFromFile (std::string filename, imageType::imageType type = \fCimageType::display\fP)"
loads image from file
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/17
.RE
.PP
FullName \fBtargeterImage::addImageFromFile\fP Qualifier 
.PP
\fBParameters:\fP
.RS 4
\fIstd::string\fP filename 
.br
\fIimageType::imageType\fP type 
.RE
.PP
\fBReturns:\fP
.RS 4
void Access public 
.RE
.PP

.PP
Definition at line 256 of file targeterimage\&.cpp\&.
.SS "void targeterImage::createQTImage (bool bRGBSwap = \fCtrue\fP)"
Creates a QImage for display purposes from an OpenCV image (shares memory)
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/13
.RE
.PP
FullName targeterImage::getQImage Qualifier 
.PP
\fBParameters:\fP
.RS 4
\fIbool\fP bRGBSwap 
.RE
.PP
\fBReturns:\fP
.RS 4
void Access private 
.RE
.PP

.PP
Definition at line 196 of file targeterimage\&.cpp\&.
.SS "void targeterImage::free1DImages ()"
free's memory used by image pointers
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/17
.RE
.PP
FullName \fBtargeterImage::free1DImages\fP Qualifier 
.PP
\fBReturns:\fP
.RS 4
void Access public 
.RE
.PP

.PP
Definition at line 130 of file targeterimage\&.cpp\&.
.SS "int * targeterImage::get1DImage (imageType::imageType type)"
Returns int* image pointer depending on type of image
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/17
.RE
.PP
FullName \fBtargeterImage::get1DImage\fP Qualifier 
.PP
\fBParameters:\fP
.RS 4
\fIimageType::imageType\fP type 
.RE
.PP
\fBReturns:\fP
.RS 4
int* Access public 
.RE
.PP

.PP
Definition at line 90 of file targeterimage\&.cpp\&.
.SS "cv::Mat & targeterImage::getImage ()"
returns the current display image if existing or a valid image if not
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/15
.RE
.PP
FullName \fBtargeterImage::getImage\fP Qualifier 
.PP
\fBReturns:\fP
.RS 4
cv::Mat& Access public 
.RE
.PP

.PP
Definition at line 216 of file targeterimage\&.cpp\&.
.SS "\fBtargeterImage\fP & targeterImage::getImageOfType (imageType::imageType imageFunction)"
gets associated image of a certain type
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/17
.RE
.PP
FullName \fBtargeterImage::getImageOfType\fP Qualifier 
.PP
\fBParameters:\fP
.RS 4
\fIimageType::imageType\fP imageFunction 
.RE
.PP
\fBReturns:\fP
.RS 4
\fBtargeterImage::targeterImage\fP& Access public 
.RE
.PP

.PP
Definition at line 156 of file targeterimage\&.cpp\&.
.SS "QImage targeterImage::getQTImage ()"
gets QImage and if not exising creates one from valid cv::Mat image
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/15
.RE
.PP
FullName \fBtargeterImage::getQTImage\fP Qualifier 
.PP
\fBReturns:\fP
.RS 4
QT_NAMESPACE::QImage Access public 
.RE
.PP

.PP
Definition at line 178 of file targeterimage\&.cpp\&.
.SS "void targeterImage::removeDerivedIndex (int ind)"
remove associated image given its ID
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/17
.RE
.PP
FullName \fBtargeterImage::removeDerivedIndex\fP Qualifier 
.PP
\fBParameters:\fP
.RS 4
\fIint\fP ind 
.RE
.PP
\fBReturns:\fP
.RS 4
void Access public 
.RE
.PP

.PP
Definition at line 365 of file targeterimage\&.cpp\&.
.SS "void targeterImage::removeDerivedIndexes (const std::vector< int > & to_remove)"
remove associated images given their ID's
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/17
.RE
.PP
FullName \fBtargeterImage::removeDerivedIndexes\fP Qualifier 
.PP
\fBParameters:\fP
.RS 4
\fIconst\fP std::vector<int> & to_remove 
.RE
.PP
\fBReturns:\fP
.RS 4
void Access public 
.RE
.PP

.PP
Definition at line 384 of file targeterimage\&.cpp\&.
.SS "void targeterImage::set1DImage (int * im, imageType::imageType type = \fCimageType::display\fP)"
sets int* image pointer, depending on type of image
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/17
.RE
.PP
FullName \fBtargeterImage::set1DImage\fP Qualifier 
.PP
\fBParameters:\fP
.RS 4
\fIint\fP * im 
.br
\fIimageType::imageType\fP type 
.RE
.PP
\fBReturns:\fP
.RS 4
void Access public 
.RE
.PP

.PP
Definition at line 111 of file targeterimage\&.cpp\&.
.SS "std::string targeterImage::setDefaultName ()"
sets a default name for image based on type
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/17
.RE
.PP
FullName \fBtargeterImage::setDefaultName\fP Qualifier 
.PP
\fBReturns:\fP
.RS 4
std::string Access public 
.RE
.PP

.PP
Definition at line 310 of file targeterimage\&.cpp\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for targeter from the source code\&.


targeterimage.cpp

//#include <vld.h>

#include "globals.h"
#include "targeterimage.h"
#include "imagesContainer.h"
#include "HelperFunctions.h"

#include <QRect>
#include <QDebug>
#include <QImage>

/**
*
*  Constructor for targeterImage class, has pointer to image array
*
* @author    David Watts
* @since     2017/03/17
* 
* FullName   targeterImage::targeterImage
* Qualifier 
* @param     ImagesContainer * pImages 
* @return    
* Access     public 
*/
targeterImage::targeterImage(ImagesContainer* pImages)
{ 
	this->m_ImagesContainer = pImages;
	pImage = NULL;
	pMask = NULL;
	imageFunction = imageType::display;
	filepathname = "";
	name = "";
	tooltip = "";
	isDisplayed = false;
}

/**
*
*   Constructor for targeterImage class
*
* @author    David Watts
* @since     2017/03/17
* 
* FullName   targeterImage::targeterImage
* Qualifier 
* @return    
* Access     public 
*/
targeterImage::targeterImage()
{
	pImage = NULL;
	pMask = NULL;
	imageFunction = imageType::display;
	filepathname = "";
	name = "";
	tooltip = "";
	isDisplayed = false;
}

/**
*
* targeterImage destructor 
*
* @author    David Watts
* @since     2017/03/17
* 
* FullName   targeterImage::~targeterImage
* Qualifier 
* @return    
* Access     public 
*/
targeterImage::~targeterImage() {
	//free1DImages(); 
}

/**
*
*  Returns int* image pointer depending on type of image
*
* @author    David Watts
* @since     2017/03/17
* 
* FullName   targeterImage::get1DImage
* Qualifier 
* @param     imageType::imageType type 
* @return    int*
* Access     public 
*/
int* targeterImage::get1DImage(imageType::imageType type) {
	if (type == imageType::mask)
		return pMask;
	else
		return pImage;
};

/**
*
*  sets int* image pointer, depending on type of image
*
* @author    David Watts
* @since     2017/03/17
* 
* FullName   targeterImage::set1DImage
* Qualifier 
* @param     int * im 
* @param     imageType::imageType type 
* @return    void
* Access     public 
*/
void targeterImage::set1DImage(int* im, imageType::imageType type) {
	if (type == imageType::mask)
		pMask = im;
	else
		pImage = im;
};

/**
*
* free's memory used by image pointers 
*
* @author    David Watts
* @since     2017/03/17
* 
* FullName   targeterImage::free1DImages
* Qualifier 
* @return    void
* Access     public 
*/
void targeterImage::free1DImages() {
	if (pImage)
	{
		delete[] pImage;
		pImage = NULL;
	}
	if (pMask)
	{
		delete[] pMask;
		pMask = NULL;
	}
};

/**
*
* gets associated image of a certain type  
*
* @author    David Watts
* @since     2017/03/17
* 
* FullName   targeterImage::getImageOfType
* Qualifier 
* @param     imageType::imageType imageFunction 
* @return    targeterImage::targeterImage&
* Access     public 
*/
targeterImage& targeterImage::getImageOfType(imageType::imageType imageFunction) {
	for (int i = 0; i < m_derivedImages.size(); i++)
	{
		if (m_ImagesContainer->getImageAt(m_derivedImages[i]).imageFunction == imageFunction)
		{ 
			return m_ImagesContainer->getImageAt(m_derivedImages[i]);
		}
	}
}

/**
*
*  gets QImage and if not exising creates one from valid cv::Mat image
*
* @author    David Watts
* @since     2017/03/15
*
* FullName   targeterImage::getQTImage
* Qualifier
* @return    QT_NAMESPACE::QImage
* Access     public
*/
QImage targeterImage::getQTImage()
{ 
	return qt_im; 
};

/**
*
* Creates a QImage for display purposes from an OpenCV image (shares memory)
*
* @author    David Watts
* @since     2017/03/13
*
* FullName   targeterImage::getQImage
* Qualifier
* @param     bool bRGBSwap
* @return    void
* Access     private
*/
void targeterImage::createQTImage(bool bRGBSwap)
{
	qt_im = QImage(cv_im.data, cv_im.cols, cv_im.rows, cv_im.step1(), HelperFunctions::getFormat(cv_im.type()));

	if (bRGBSwap)
		qt_im = qt_im.rgbSwapped();
}

/**
*
*  returns the current display image if existing or a valid image if not
*
* @author    David Watts
* @since     2017/03/15
* 
* FullName   targeterImage::getImage
* Qualifier 
* @return    cv::Mat&
* Access     public 
*/
cv::Mat& targeterImage::getImage()
{
	// always try to return a valid image
	return cv_im;
}

int targeterImage::getImageRows()
{
	return getImage().rows;
}

int targeterImage::getImageCols()
{
	return getImage().cols;
}

int targeterImage::getImageType()
{
	return getImage().type();
}


cv::Size targeterImage::getImageSize() {
	return getImage().size();
}

/**
*
*  loads image from file
*
* @author    David Watts
* @since     2017/03/17
* 
* FullName   targeterImage::addImageFromFile
* Qualifier 
* @param     std::string filename 
* @param     imageType::imageType type 
* @return    void
* Access     public 
*/
void targeterImage::addImageFromFile(std::string filename, imageType::imageType type)
{
	cv::Mat im = cv::imread(filename);

	this->filepathname = filename;

	this->addImageEx(im, type, filename);
}

/**
*
*  Extended method to create image
*
* @author    David Watts
* @since     2017/03/17
* 
* FullName   targeterImage::addImageEx
* Qualifier 
* @param     cv::Mat im 
* @param     imageType::imageType type 
* @param     std::string descriptiveName 
* @param     std::string tooltip 
* @param     std::string filename 
* @return    void
* Access     public 
*/
void targeterImage::addImageEx(cv::Mat im, imageType::imageType type, std::string descriptiveName, std::string tooltip, std::string filename)
{
	this->name = descriptiveName;
	this->imageFunction = type;

	this->filepathname = filename;

	addImage(im);

	if (descriptiveName == "")
		this->name = setDefaultName();

	if(tooltip == "")
		this->tooltip = this->name;
}

/**
*
*  sets a default name for image based on type
*
* @author    David Watts
* @since     2017/03/17
* 
* FullName   targeterImage::setDefaultName
* Qualifier 
* @return    std::string
* Access     public 
*/
std::string targeterImage::setDefaultName()
{
	if (this->imageFunction == imageType::display)
		return "an image";
	else if (this->imageFunction == imageType::mask)
		return "mask image";
	else if (this->imageFunction == imageType::cclabels)
		return "connected components image";
	else if (this->imageFunction == imageType::target)
		return "target image";
	else if (this->imageFunction == imageType::test)
		return "detection image";
	else if (this->imageFunction == imageType::score)
		return "score image";
	else if (this->imageFunction == imageType::score)
		return "centroids image";
	else if (this->imageFunction == imageType::roi)
		return "region of interest image";
	else
		return "an image";
}

/**
*
*  adds image and creates QT Image copy for display
*
* @author    David Watts
* @since     2017/03/17
* 
* FullName   targeterImage::addImage
* Qualifier 
* @param     cv::Mat im 
* @return    void
* Access     public 
*/
void targeterImage::addImage(cv::Mat im)
{
	this->cv_im = im;

	createQTImage(true);
}

/**
*
*  remove associated image given its ID
*
* @author    David Watts
* @since     2017/03/17
* 
* FullName   targeterImage::removeDerivedIndex
* Qualifier 
* @param     int ind 
* @return    void
* Access     public 
*/
void targeterImage::removeDerivedIndex(int ind) {
	std::vector<int>::iterator it = m_derivedImages.begin();
	std::advance(it, ind);
	m_derivedImages.erase(it);
}

/**
*
*  remove associated images given their ID's
*
* @author    David Watts
* @since     2017/03/17
* 
* FullName   targeterImage::removeDerivedIndexes
* Qualifier 
* @param     const std::vector<int> & to_remove 
* @return    void
* Access     public 
*/
void targeterImage::removeDerivedIndexes(const std::vector<int>& to_remove)
{
	auto vector_base = m_derivedImages.begin();
	std::vector<int>::size_type down_by = 0;

	for (auto iter = to_remove.cbegin();
		iter < to_remove.cend();
		iter++, down_by++)
	{
		std::vector<int>::size_type next = (iter + 1 == to_remove.cend()
			? m_derivedImages.size()
			: *(iter + 1));

		std::move(vector_base + *iter + 1,
			vector_base + next,
			vector_base + *iter - down_by);
	}
	m_derivedImages.resize(m_derivedImages.size() - to_remove.size());
}

